core/src/main/scala/kafka/tools/ConsoleProducer.scala

```diff
@@ -264,39 +278,92 @@ object ConsoleProducer {
+      if (headersDelimiter.equals(headersSeparator))
+        throw new KafkaException("headers.delimiter and headers.separator may not be equal")
+      if (headersDelimiter.equals(headerKeySeparator))
+        throw new KafkaException("headers.delimiter and headers.key.separator may not be equal")
+      if (headersSeparator.equals(headerKeySeparator))
+        throw new KafkaException("headers.separator and headers.key.separator may not be equal")
       reader = new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8))
     }
 
-    override def readMessage() = {
+    override def readMessage(): ProducerRecord[Array[Byte], Array[Byte]] = {
       lineNumber += 1
-      if (printPrompt)
-        print(">")
-      (reader.readLine(), parseKey) match {
-        case (null, _) => null
-        case (line, true) =>
-          line.indexOf(keySeparator) match {
-            case -1 =>
-              if (ignoreError) new ProducerRecord(topic, line.getBytes(StandardCharsets.UTF_8))
-              else throw new KafkaException(s"No key found on line $lineNumber: $line")
-            case n =>
-              val value = (if (n + keySeparator.size > line.size) "" else line.substring(n + keySeparator.size)).getBytes(StandardCharsets.UTF_8)
-              new ProducerRecord(topic, line.substring(0, n).getBytes(StandardCharsets.UTF_8), value)
+      if (printPrompt) print(">")
+      val line = reader.readLine()
+      line match {
+        case null => null
+        case line =>
+
+          val headers = parse(parseHeader, line, headersDelimiter, "headers delimiter")
+          val key = parse(parseKey, line.substring(offset(headers)), keySeparator, "key separator")
+          val value = line.substring(offset(headers) + offset(key))
+
+          val record = new ProducerRecord[Array[Byte], Array[Byte]](
+            topic,
+            if (key != null) key.getBytes(StandardCharsets.UTF_8) else null,
+            if (value != null) value.getBytes(StandardCharsets.UTF_8) else null,
+          )
+
+          if (headers != null){
+            splitHeaders(headers)
+              .foreach(header => record.headers().add(header._1, header._2))
           }
-        case (line, false) =>
-          new ProducerRecord(topic, line.getBytes(StandardCharsets.UTF_8))
+
+          record
       }
     }
+
+    private def parse(enabled: Boolean, toParse: String, demarcation: String, demarcationName: String): String = {
+      (enabled, toParse.indexOf(demarcation)) match {
+        case (false, _) => null
+        case (_, -1) =>
+          if (ignoreError) null
+          else throw new KafkaException(s"No $demarcationName found in '$toParse' on line number $lineNumber")
+        case (_, index) => toParse.substring(0, index)
+      }
+    }
+
+    private def splitHeaders(headers: String): Array[(String, Array[Byte])] = {
+      headerSeparatorPattern.split(headers)
+        .map(pair =>
```

Comment:
nit: Could we move `map` to the previous line and use curly braces instead of parenthesis? That would be more align with our code style.